elyra/pipeline/component_registry.py

```diff
@@ -0,0 +1,271 @@
+
+    @property
+    def registry_location(self) -> str:
+        return self._component_registry_location
+
+    def get_all_components(self) -> List[Component]:
+        """
+        Retrieve all components from the component_id registry
+        """
+        components: List[Component] = list()
+
+        # Read component_id catalog to get JSON
+        component_entries = self._read_component_registry()
+
+        for component_id, component_entry in component_entries.items():
+            self.log.debug(f"Component registry found component_id {component_entry.get('name')}")
+
+            # Get appropriate reader to read component_id definition
+            reader = self._get_reader(component_entry)
+
+            component_location = component_entry['location'][reader._type]
+            # Adjust location to dislay full path
+            if reader._type == "filename":
+                component_location = os.path.join(os.path.dirname(__file__), component_location)
+
+            component_definition = reader.read_component_definition(component_location, component_id)
+
+            # Parse component_id details and add to list
+            component = self._parser.parse(component_id, component_definition)
+            if component:
+                components.extend(component)
+
+        return components
+
+    def get_component_catalog_entry(self, component_id):
+        """
+        Get the body of the component_id catalog entry with the given id
+        """
+        # Read component_id catalog to get JSON
+        component_entries = self._read_component_registry()
+
+        # Find entry with the appropriate id, if exists
+        component_entry = next((entry for id, entry in component_entries.items() if component_id == id), None)
+        if not component_entry:
+            self.log.error(f"Component with ID '{component_id}' could not be found in the " +
+                           f"{self._component_registry_location} component_id catalog.")
+            raise ValueError(f"Component with ID '{component_id}' could not be found in the " +
+                             f"{self._component_registry_location} component_id catalog.")
+
+        return component_entry
+
+    def get_component(self, component_id):
+        """
+        Return the properties JSON for a given component_id.
+        """
+        # Read component catalog to get component with given id
+        adjusted_id = self._parser.get_adjusted_component_id(component_id)
+        component_entry = self.get_component_catalog_entry(adjusted_id)
+
+        # Get appropriate reader to read component_id definition
+        reader = self._get_reader(component_entry)
+
+        component_location = component_entry['location'][reader._type]
+        # Adjust location to dislay full path
+        if reader._type == "filename":
+            component_location = os.path.join(os.path.dirname(__file__), component_location)
```

Comment:
Are we going to have all file-based component definitions within our installation area?

Also, since `component_location` is already a function of `reader._type` (3 lines above), can we perform this additional prefixing when `component_entry['location']['filename']` is getting initialized and avoid this adjustment here?

Reply:
I think long term we probably don’t want components stored here, so you’re right to question it. 

I like the idea of doing the prefixing in init since it doesn’t fit very well! Then we can remove the duplicate code and still have the full file path displayed when passing parameters to the front end.

Reply:
I believe by housing component definitions within the installation area, any additional definitions (added post-installation) will be removed when uninstalled.  Given these probably have no use elsewhere, that may not be a big deal, but it's a bit of a nasty side-effect that users should be aware of.  If this approach is retained, it might be better to ask users to contribute their definitions - making them part of our release (which also could be nasty). :smile:

I agree with just moving forward with this approach for now and perhaps see what the "involvement" is like.